Added an example of per-project dependency management tool. Fixes #162. (#336)
carpentries-incubator/python-intermediate-development: _episodes/12-virtual-environments.md

plus a number of additional external libraries.

Virtual environments are not just a feature of Python -

> most modern programming languages use them to isolate libraries/dependencies for a specific project

Reviewer: ```suggestion
most modern programming languages use them to isolate libraries or dependencies for a specific project
```

and make it easier to develop, run, test and share code with others.
Even languages that do not explicitly have virtual environments have mechanisms

> that promote per-project dependency collections (e.g. Maven with classpath for Java).

Reviewer: Might be nice to give some more examples here? I think I included more in the slides that I use, but just using those from the linked issue #162 : 

```suggestion
that promote per-project dependency collections:

- Maven with classpath for Java
- Bundler for Ruby
- Conan and per-project environments for C++
- Spack, used extensively for package management in HPC contexts, is a more generic, language agnostic tool that also has something resembling virtual environments
```

Author: Do not know about Conan and Spack but Bundler for ruby is actually a tool that enables creation of virtual environments similar to venv (locking the ruby and gem versions) and we are trying to give examples of tools that are not explicitly doing that but are providing some kind of dependency management. For that reason I did not include any of the above even though they were mentioned in the issue. Perhaps a good example does not exist? Maybe we just need to get rid of that sentence: "Even languages that don't explicitly have virtual environments have other mechanisms that promote per-project library collections." @bielsnohr ?

Reviewer: I think the confusing bit is that most languages combine the tooling that does package management and dependency isolation, whereas in Python they are distinct tools that can work together (pip and venv). I agree that we should probably just get rid of the sentence

> Even languages that do not explicitly have virtual environments have mechanisms that promote per-project dependency collections (e.g. Maven with classpath for Java).

"Virtual environments" is a term very specific to Python, and it doesn't really apply well to other languages. I'll suggest an alternate wording in another review comment.

plus a number of additional external libraries.

> Virtual environments are not just a feature of Python -

Reviewer: ```suggestion
Virtual environments are not just a feature of Python---
```

Author: I like this proposal - think it explains things more clearly - thank you @bielsnohr.

most modern programming languages use them to isolate libraries or dependencies for a specific project
and make it easier to develop, run, test and share code with others.
Even languages that do not explicitly have virtual environments have mechanisms

> that promote per-project dependency collections (e.g. Maven with classpath for Java).

Author: Accepted changes - you still need to approve the PR @bielsnohr.
